homeassistant/components/zwave_js/api.py

```diff
@@ -786,3 +796,129 @@ async def get(self, request: web.Request, config_entry_id: str) -> web.Response:
+    hass: HomeAssistant,
+    connection: ActiveConnection,
+    msg: dict,
+    node: Node,
+) -> None:
+    """Subsribe to the status of a firmware update."""
+
+    @callback
+    def async_cleanup() -> None:
+        """Remove signal listeners."""
+        for unsub in unsubs:
+            unsub()
+
+    @callback
+    def forward_progress(event: dict) -> None:
+        progress: FirmwareUpdateProgress = event["firmware_update_progress"]
+        connection.send_message(
+            websocket_api.event_message(
+                msg[ID],
+                {
+                    "sent_fragments": progress.sent_fragments,
+                    "total_fragments": progress.total_fragments,
+                },
+            )
+        )
+
+    @callback
+    def forward_finished(event: dict) -> None:
+        finished: FirmwareUpdateFinished = event["firmware_update_finished"]
+        connection.send_message(
+            websocket_api.event_message(
+                msg[ID],
+                {
+                    "status": finished.status,
+                    "wait_time": finished.wait_time,
+                },
+            )
+        )
+
+    unsubs = [
+        node.on("firmware update progress", forward_progress),
+        node.on("firmware update finished", forward_finished),
+    ]
+    connection.subscriptions[msg["id"]] = async_cleanup
+
+    connection.send_result(msg[ID])
+
+
+class FirmwareUploadView(HomeAssistantView):
+    """View to upload firmware."""
+
+    url = "/api/zwave_js/firmware/upload/{config_entry_id}/{node_id}"
+    name = "api:zwave_js:firmware:upload"
+
+    async def post(
+        self, request: web.Request, config_entry_id: str, node_id: str
+    ) -> web.Response:
+        """Handle upload."""
+        hass = request.app["hass"]
+        if config_entry_id not in hass.data[DOMAIN]:
+            raise web_exceptions.HTTPBadRequest
+
+        entry = hass.config_entries.async_get_entry(config_entry_id)
+        client = hass.data[DOMAIN][config_entry_id][DATA_CLIENT]
+        node = client.driver.controller.nodes.get(int(node_id))
+        if not node:
+            raise web_exceptions.HTTPBadRequest
+
+        # Increase max payload
+        request._client_max_size = 1024 * 1024 * 10  # pylint: disable=protected-access
+
+        data = await request.post()
+
+        if "file" not in data or not isinstance(data["file"], web_request.FileField):
+            raise web_exceptions.HTTPBadRequest
+
+        uploaded_file: web_request.FileField = data["file"]
+
+        status = await begin_firmware_update(
+            entry.data[CONF_URL],
+            node,
+            uploaded_file.filename,
+            await hass.async_add_executor_job(uploaded_file.file.read),
+            async_get_clientsession(hass),
+        )
+
+        if not status["success"]:
+            raise web_exceptions.HTTPUnprocessableEntity(
+                reason=f"{status['zwaveErrorMessage']} ({status['zwaveErrorCode']})"
```

Comment:
Where is this reason printed?

Reply:
I will be reverting this piece of the PR, I learned that the client will raise if an error is received